### DISCUSSION TITLE
Subscription in OPCUA Server

### DISCUSSION THREAD

**Discussion by vidyarao123** [FreeOpcUa/python-opcua#1239] [Q&A]
I am trying to add subscription in my OPCUA Server code but the keyword "create_subscription" ( the one as per the OPCUA docs) seems to be not being recognized by the code. 

Below is a part of my code where I am trying to add subscription

name = "PYTHON_OPCUA_SERVER"
addspace = server.register_namespace(name)
node = server.get_objects_node()
Param = node.add_object(addspace, "Parameters")
Temp = Param.add_variable(addspace, "Temperature", 0) 
server.start()
Temp.set_writable()
Temp.create_subscription()

1. Can somebody explain how to **create subscription** and use the **subscribe_data_change** functionality  in OPCUA server code ?
2. I saw the subscription.py in https://github.com/FreeOpcUa/python-opcua/blob/master/opcua/common/subscription.py . Am I supposed to copy the entire py file in order to access the subscription class?

I am an amateur when it comes to python with OPCUA . Kindly bear with me :) 
Thanks

**Comment by swamper123** [comment]
Ahoi,

for subscriptions I recommened the [example file](https://github.com/FreeOpcUa/python-opcua/blob/master/examples/server-example.py) for a better "quick start" in that topic.

With **create_subscription(params, handler)** you create a Subscription object and telling the server "I want to subscribe something, which should be checked every _period/params_ milliseconds and how should be deald with it shall be organised in _handler_.

A handler can be like this one:

```python
class SubHandler(object):
    """
    Subscription Handler. To receive events from server for a subscription
    """

    def datachange_notification(self, node, val, data):
        print("Python: New data change event", node, val)

    def event_notification(self, event):
        print("Python: New event", event)


[...add here standard main code like you did ...]

params = 500 # One possible option, for more details see in sourcecode
sub = client.create_subscription(params, Subhandler)

```
but you can add other handlers, more methods, depending on your needs. So now you told the server "This is one way, how I want to deal a Subscription". But what shall the server looking for? Which Node shall be subscribed now?


With **sub.subscribe_data_change(mynode)** you say: "I want to Subscribe the node _mynode_, which should be handled like mentioned in the Subscriptionobject _sub_ above, to an _datachange_notification_ event. If such an event happens, do whatever is inside the method which you find in the Subhandler() class.

So your code may have to change a bit (no guaranty):

```python 
 class SubHandler(object):
    """
    Subscription Handler. To receive events from server for a subscription.
    Never add long running code inside these methods! They are blocking the server otherwise!
    """

    def datachange_notification(self, node, val, data):
        print("Python: New data change event", node, val)

    def event_notification(self, event):
        print("Python: New event", event)


name = "PYTHON_OPCUA_SERVER"
addspace = server.register_namespace(name)
node = server.get_objects_node()
Param = node.add_object(addspace, "Parameters")
Temp = Param.add_variable(addspace, "Temperature", 0)
Temp.set_writable()
handler = SubHandler()
sub = server.create_subscription(500, datachange_notification) # example!
handle = sub.subscribing_data_cahnge(Temp) # sub to value changes of the node Temp
server.start()

# just an ugly out-of-the-box idea to trigger datachange events.
for i in 1000:
    Temp.set_value(i)

```

In case of faults, please correct me.

**Reply by ratara** [reply]
Thanks alot. This is the missing documentation.
Maybe such things should be documented either in the code or as external api documentation.

By the way. The handler would be more self-documenting, if there would be a abstract interface for it.

EDIT: After ready the example again (RTFM), i noticed why i am getting errors on internal variables being None. Subscription can only be created when the server was started before. Server-side, this was not clear to me.

**Comment by vidyarao123** [comment]
Hi @swamper123
Your response on subscription concept was pretty clear. And its working fine now. 
I did a small change though

class SubHandler(object):
    """
    Subscription Handler. To receive events from server for a subscription.
    Never add long running code inside these methods! They are blocking the server otherwise!
    """

    def datachange_notification(self, node, val, data):
        print("Python: New data change event", node, val)

    def event_notification(self, event):
        print("Python: New event", event)
        
name = "PYTHON_OPCUA_SERVER"
addspace = server.register_namespace(name)
node = server.get_objects_node()
Param = node.add_object(addspace, "Parameters")
Temp = Param.add_variable(addspace, "Temperature", 0) 

Temp.set_writable()
server.start()
handler = SubHandler()
**sub = server.create_subscription(500, handler)** # example!
handle = sub.subscribe_data_change(Temp) # sub to value changes of the node Temp 
print (handle)

In the highlighted line of the above code, I changed the argument of create_subscription to "handler" as I was getting an error when used datachange_notification. The rest are the same. 

I have another question. 
Now that I am getting notifications on the server side, which makes me wonder, if there is a way to get a notification, when the client writes back to the same variable (Basically in Read & Write case ) 

Your help is much appreciated ..
Thanks :)

**Reply by swamper123** [reply]
Ahoi @vidyarao123 !

have you got a` SubHandler()` class like me? Because that is needed WITH a method called `datachange_notification()`.

Then you should get an event triggered, if the value, you subscribed to, was changed (in an OPC UA way).

**Reply by vidyarao123** [reply]
Yes the class is the same as yours

**Reply by swamper123** [reply]
\* browses throw lib \* ahm... ups. :D 
Your solution seems to be right.

Do you mean to get a notification on Client side, if a value was changed (example the client writes a new value)? 
Or a notification on server side ("value was changed?" - which should be possible now, if you got the SubHandler class like me?!)?
What kind of events do you get otherwise?

**Reply by vidyarao123** [reply]
Well, I am getting a notification, whenever there is a change in the event. But I am interested to get a notification especially when the value is written from OPC Client.

**Reply by vidyarao123** [reply]
I mean the notification on server side, when client writes the value back to server. Right now I am getting a notification for every change. But I want the notification only when it's written from the client side. I hope I am clear . :)

**Reply by vidyarao123** [reply]
The subhandler class is like yours ..

**Reply by AndreasHeine** [reply]
tbh thats not how a opcua subscriptions work! whenever the subscribed value changes all subscribers will be notified regardless who changed the value. https://reference.opcfoundation.org/v104/Core/docs/Part4/5.13.1/
